datafusion/physical-expr/src/aggregate/sum.rs

```diff
@@ -101,6 +106,43 @@ impl AggregateExpr for Sum {
     }
 
     fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
+        if let DataType::List(field) = &self.data_type {
```

Comment:
@alamb I agree that we should not introduce any downgrade especially for this widely used function.

I think we can either run the comparison that ensure there is no performance downgrade for `sum` or we just consider the initial approach that we build `array_aggreate` like other array function.

In current approach, ArraySum is actually very different from Sum. We have another Accumulator for Array version. I think this is the only place that might effect the performance of current sum aggregate, where we need to differentiate Array and non-Array cases. 

Actually, the initial goal that we try to done this in Accumulator "To reduce code duplication" is actually no longer true. Unless we have other reason that we need ArraySum Accumulator. Maybe we should move this back to `array_expression.rs`?

1. Is comparison the performance of Sum enough to ensure there should be no downgrade?
2. Should we have ArraySum Accumulator for other reason or feature?
3. If we implement array aggregate function in `array_expression.rs` is there any concern for the overall design?

Reply:
> If we implement array aggregate function in array_expression.rs is there any concern for the overall design?

Maybe we could put it in array_aggregates.rs (array_expressions.rs is already quite substantial)

Upon further reflection I agree that it makes sense to keep ArraySum separate from the normal Sum aggregator